### PR TITLE
MDEV-28263: mariadb-tzinfo-to-sql improve wsrep and binlog cases

### PR DESCRIPTION
**Pull request by grooverdan**
- [x] *The Jira issue number for this PR is: MDEV-28263*

## Description

The --skip-write-binlog message was confusing that it only had
an effect if the galera was enabled. There are uses beyond galera
so we apply SET SESSION SQL_LOG_BIN=0 as implied by the option
without being conditional on the wsrep status.

We also with --skip-write-binlog actually check the session @@WSREP_ON
variable rather than the global server variable.

Since 10.6, the wsrep_mode could replicate Aria, in which case no change
to innodb and back is needed.

By removing the conditions, we can use LOCK TABLES in a general case
improving the load speed of Aria (MDEV-23326), regardless of the
skip-write-binlog flag. The only case where we don't use LOCK TABLES is
when we are replicating via Innodb, because wsrep_on=1 and wsrep_mode
doesn't contain REPLICATE_ARIA. This uses an Innodb transaction instead.

By removing the \d and other syntax that requires parsing by
the mariadb client, we can use the generated SQL more generally, like
in the embedded server.

We also save and restore the SQL_LOG_BIN and WSREP_ON session server
variables so this can be included in the same session as other data
without taking into changes in state.

Remove wsrep.mysql_tzinfo_to_sql_symlink{,_skip} tests as they offered
no additional coverage beyond main.mysql_tzinfo_to_sql_symlink (no
server testing was done).

Add galera.mariadb_tzinfo_to_sql to actually test the replication
of tzinfo data through galera.

The conditional executable comment around /*M!100602 .. LOCK TABLES.. */
is so that we can provide tzinfo files (MDEV-27113, MDBF-389) and in
the case that a user uses it on a pre-10.6 server version it will still work.
LOCK TABLES isn't supported in prepared statements earlier than 10.6.2.

## How can this PR be tested?

TODO: modify the automated test suite to verify that the PR causes MariaDB to
behave as intended. Consult the documentation on
["Writing good test cases"](https://mariadb.org/get-involved/getting-started-for-developers/writing-good-test-cases-mariadb-server).
In many cases, this will be as simple as modifying one `.test` and one `.result`
file in the `mysql-test/` subdirectory. Without _automated_ tests, future regressions
in the expected behavior can't be automatically detected and verified.

If the changes are not amenable to automated testing, please explain why not and
carefully describe how to test manually.

## Basing the PR against the correct MariaDB version
- [ ] *This is a new feature and the PR is based against the latest MariaDB development branch*
- [ X ] *This is a bug fix and the PR is based against the earliest branch in which the bug can be reproduced*

The earliest version that is easy to implement/test.

### CODE REVIEW

File: mysql-test/suite/galera/t/mariadb_tzinfo_to_sql.test [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+CREATE TABLE time_zone_leap_second LIKE time_zone_leap_second_orig;
+
+SET @save_wsrep_mode=@@WSREP_MODE;
+
+--echo #
+--echo # Run on zoneinfo directory --skip-write-binlog
+--echo #
+
+--replace_result $MYSQLTEST_VARDIR MYSQLTEST_VARDIR
+--exec $MYSQL_TZINFO_TO_SQL --skip-write-binlog $MYSQLTEST_VARDIR/zoneinfo 2>&1
```
Comment (bnestere) [inline mysql-test/suite/galera/t/mariadb_tzinfo_to_sql.test:35]: Do you need to write the SQL output to stdout here? I feel that could cause some hardship with future additions (besides `mysql_tzinfo_to_sql_symlink.test` already does it).

File: sql/tztime.cc [hunk 2]
```diff
@@ -2832,17 +2821,16 @@ main(int argc, char **argv)
...
-      "ALTER TABLE time_zone_transition_type ENGINE=Aria, ORDER BY Time_zone_id, Transition_type_id;\n"
-      "END IF|\n"
-      "\\d ;\n");
+  if(opt_skip_write_binlog)
+    printf("SET SESSION SQL_LOG_BIN=@save_sql_log_bin;\n"
+           "execute immediate if(@wsrep_is_on, 'SET SESSION WSREP_ON=@save_wsrep_on', 'do 0');\n");
+  else
+    // Change back to Aria
+    printf(
+      "execute immediate if(@wsrep_cannot_replicate_tz, 'ALTER TABLE time_zone ENGINE=Aria', 'do 0');\n"
```
Comment (bnestere) [inline sql/tztime.cc:2830]: Hmm, I see it was like this before, but are the `time_zone` tables always initially guaranteed to be `Aria`  (now and in future releases)? Would it make sense to save their current engine at script start, and revert to it here? (using the same pattern as `@save_sql_log_bin`)?

**Review by bnestere — COMMENTED** [review]
Left a few questions, generally looks good though.

File: mysql-test/suite/galera/t/mariadb_tzinfo_to_sql.test [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+--exec mkdir $MYSQLTEST_VARDIR/zoneinfo
+--exec ln -s $MYSQLTEST_VARDIR/zoneinfo $MYSQLTEST_VARDIR/zoneinfo/posix
+--copy_file std_data/zoneinfo/GMT $MYSQLTEST_VARDIR/zoneinfo/GMT
+
+--echo
+--echo # On node_1
+--connection node_1
+
+use mysql;
+RENAME TABLE time_zone TO time_zone_orig,
```
[the remark this answers is not in the document]
Comment (grooverdan) [reply mysql-test/suite/galera/t/mariadb_tzinfo_to_sql.test:17]: It makes the test case harder to read as these tables are partially populated:
```
CURRENT_TEST: galera.mariadb_tzinfo_to_sql
--- /home/dan/repos/mariadb-server-10.6/mysql-test/suite/galera/r/mariadb_tzinfo_to_sql.result	2022-04-19 17:03:14.198741098 +1000
+++ /home/dan/repos/mariadb-server-10.6/mysql-test/suite/galera/r/mariadb_tzinfo_to_sql.reject	2022-04-20 07:38:34.109889029 +1000
@@ -7,16 +7,11 @@
 # On node_1
 connection node_1;
 use mysql;
-RENAME TABLE time_zone TO time_zone_orig,
-time_zone_name TO time_zone_name_orig,
-time_zone_transition TO time_zone_transition_orig,
-time_zone_transition_type TO time_zone_transition_type_orig,
-time_zone_leap_second TO time_zone_leap_second_orig;
-CREATE TABLE time_zone LIKE time_zone_orig;
-CREATE TABLE time_zone_name LIKE time_zone_name_orig;
-CREATE TABLE time_zone_transition LIKE time_zone_transition_orig;
-CREATE TABLE time_zone_transition_type LIKE time_zone_transition_type_orig;
-CREATE TABLE time_zone_leap_second LIKE time_zone_leap_second_orig;
+CREATE TABLE time_zone_orig LIKE time_zone;
+CREATE TABLE time_zone_name_orig LIKE time_zone_name;
+CREATE TABLE time_zone_transition_orig LIKE time_zone_transition;
+CREATE TABLE time_zone_transition_type_orig LIKE time_zone_transition_type;
+CREATE TABLE time_zone_leap_second_orig LIKE time_zone_leap_second;
 SET @save_wsrep_mode=@@WSREP_MODE;
 #
 # Run on zoneinfo directory --skip-write-binlog
@@ -81,26 +76,26 @@
 2
 SELECT COUNT(*) FROM time_zone_leap_second;
 COUNT(*)
-0
+22
 
 # On node_2 (not replicated)
 
 connection node_2;
 SELECT COUNT(*) FROM mysql.time_zone;
 COUNT(*)
-0
+5
 SELECT COUNT(*) FROM mysql.time_zone_name;
 COUNT(*)
-0
+6
 SELECT COUNT(*) FROM mysql.time_zone_transition;
 COUNT(*)
-0
+393
 SELECT COUNT(*) FROM mysql.time_zone_transition_type;
 COUNT(*)
-0
+31
 SELECT COUNT(*) FROM mysql.time_zone_leap_second;
 COUNT(*)
-0
+22
 
 # Apply on node_1
 
```

File: mysql-test/main/mysql_tzinfo_to_sql_symlink.test [hunk 4]
```diff
@@ -87,7 +101,16 @@ TRUNCATE TABLE time_zone_leap_second;
 --echo #
 
 --exec $MYSQL_TZINFO_TO_SQL --skip-write-binlog $MYSQLTEST_VARDIR/zoneinfo/GMT XXX 2>&1
---exec $MYSQL_TZINFO_TO_SQL --skip-write-binlog $MYSQLTEST_VARDIR/zoneinfo/GMT 2>/dev/null | $MYSQL -S $MASTER_MYSOCK -u root mysql
+--exec $MYSQL_TZINFO_TO_SQL --skip-write-binlog $MYSQLTEST_VARDIR/zoneinfo/GMT XXX 2>/dev/null > $MYSQL_TMP_DIR/tz.sql
+--disable_query_log
+--source $MYSQL_TMP_DIR/tz.sql
```
[the remark this answers is not in the document]
Comment (grooverdan) [reply mysql-test/main/mysql_tzinfo_to_sql_symlink.test:106]: Added to galera test. I didn't see other main tests that enabled binary logs.

File: sql/tztime.cc (hunk 2 above)
Comment (grooverdan) [reply sql/tztime.cc]: Implemented.

**Comment by grooverdan** [comment]
Ok, see the `HERE VARIABLE_NAME IN ('COM_BEGIN', 'COM_LOCK_TABLES', 'COM_ALTER_TABLE', 'COM_TRUNCATE')` where a mtr worker is reused stats are accumulated. Need to work that one out.

**Review by bnestere — APPROVED** [review]
Looks good! Great idea with the temp table to calculate a difference. I'll be adding that trick to my arsenal :)

**Pull request merged**